File: src/js/data/charity/Output.js

import {assert, assMatch} from 'sjtest';
import _ from 'lodash';
import DataClass, {getType} from '../../base/data/DataClass';

/* Output type, which also does Impact: 
{
	name {String}
	number: {Number} Number of units output, e.g. the number of malaria nets
	cost: {Money} total cost, ie cost = costPerOutput * number
	costPerOutput {Money}
	amount: {String} non-numerical descriptions of how much was output
	confidence {String}
	description: {String}
	start: {Date}
	end: {Date}
	order: {Number} for display lists
	year: {Number}
}
*/

/** impact utils */
class Output extends DataClass {
	/** @type {String} */
	name;
	/** @type {Number} Number of units output, e.g. the number of malaria nets */
	number;
	// cost: {Money} total cost, ie cost = costPerOutput * number
	// costPerOutput {Money}
	// amount: {String} non-numerical descriptions of how much was output
	// confidence {String}
	// description: {String}
	// start: {Date}
	// end: {Date}
	// order: {Number} for display lists
	// year: {Number}
}
DataClass.register(Output);
const This = Output;
export default Output;

// something is making outputs without a type. oh well -- also allow a duck type test for costPerBen
Output.isa = (obj) => {
	return getType(obj)==='Output' || obj.costPerBeneficiary || obj.number;
};

Output.number = obj => This.assIsa(obj) && obj.number;
Output.cost = obj => This.assIsa(obj) && obj.cost;

/**
 * A scaled version 
 * @param donationAmount {Money}
 */
Output.scaleByDonation = (output, donationAmount) => {
	// deep copy
	let impact = _.cloneDeep(output);
	// TODO scaled by donationAmount
	// TODO change units if necc
	// TODO Java needs a mirror of this :(
	console.error("scale!", impact, donationAmount);
	return impact;
};

File: src/main/java/org/sogive/data/charity/Output.java
package org.sogive.data.charity;

import java.util.Map;

import com.goodloop.data.Money;
import com.winterwell.utils.StrUtils;
import com.winterwell.web.ajax.JThing;

/**
 * This is also the type for Impact! See Output.js which documents this type
 * better
 * 
 * @author daniel
 *
 */
public class Output extends Thing<Output> {

	private static final long serialVersionUID = 1L;

	public static boolean match(Output a, Output b) {
		if (a == b)
			return true;
		if (a.getName() != null && b.getName() != null) {
			return a.getName().equals(b.getName());
		}
		return a.equals(b);
	}

	Output() {

	}

	/**
	 * 
	 * @param number How many malaria nets or nurses or kittens? Can be null for
	 *               unquantified
	 * @param type   e.g. "solar-light"
	 */
	public Output(Double number, String type) { // , String unit) {
		put("number", number);
//		put("type", StrUtils.toCanonical(type));
		put("name", type);
//		put("unit", StrUtils.toCanonical(unit));
		assert number == null || number >= 0 : this;
	}

	/**
	 * amount is for non-numerical descriptions of how much was output
	 * 
	 * @return
	 */
	public String getAmount() {
		return StrUtils.str(get("amount"));
	}

	public Money getCostPerBeneficiary() {
		Object cpb = get("costPerBeneficiary");
		if (cpb == null)
			return null;
		if (cpb instanceof Money)
			return (Money) cpb;
		return (Money) new JThing().setType(Money.class).setMap((Map) cpb).java();
	}

	/**
	 * 
	 * @return default to 1
	 */
	public double getNumber() {
		Double n = getDouble("number");
		return n==null? 1 : n;
	}

	/**
	 * A fresh output scaled by fraction. E.g. use case: If total costs were £100,
	 * and I donate £1, then scale by 0.01, to get my impact. TODO This can also
	 * change the unit, e.g. from years to days.
	 * 
	 * @param input
	 */
	public Output scale(double fraction) {
		// the number
		assert fraction >= 0 : fraction + " " + this;
		Double number = getNumber();
		Output scaled;
		if (number == null) {
			scaled = new Output(null, getName());
		} else {
			number = number * fraction;
			// String type = (String) get("type");
			// String unit = (String) getName();
			// TODO adjust the unit?
			scaled = new Output(number, getName());
		}
		// copy other properties, e.g. image
		for (Map.Entry<String, Object> pv : entrySet()) {
			if (scaled.containsKey(pv.getKey()))
				continue;
			scaled.put(pv.getKey(), pv.getValue());
		}
		return scaled;
	}

	public void setCostPerBeneficiary(Money cpb) {
		put("costPerBeneficiary", cpb);
	}

	Output setIndirect(boolean indirect) {
		put("indirect", indirect);
		return this;
	}

	public void setName(String name) {
		put("name", name);
	}

}
